runtime/flash/src/flash_storage_to_pages.rs

```diff
@@ -0,0 +1,417 @@
+            .map_or(Err(ErrorCode::RESERVE), move |page_buffer| {
+                let page_size = page_buffer.as_mut().len();
+
+                self.state.set(State::Write);
+                self.length.set(length);
+
+                if address % page_size == 0 && length >= page_size {
+                    // This write is aligned to a page and we are writing an entire page.
+                    // Copy data into page buffer.
+                    page_buffer.as_mut()[..page_size].copy_from_slice(&buffer[..page_size]);
+
+                    self.buffer.replace(buffer);
+                    self.address.set(address + page_size);
+                    self.remaining_length.set(length - page_size);
+                    self.buffer_index.set(page_size);
+
+                    match self.driver.write_page(address / page_size, page_buffer) {
+                        Ok(()) => Ok(()),
+                        Err((error_code, page_buffer)) => {
+                            self.page_buffer.replace(page_buffer);
+                            Err(error_code)
+                        }
+                    }
+                } else {
+                    // This write is non-page-aligned, so we need to do a read first.
+                    self.buffer.replace(buffer);
+                    self.address.set(address);
+                    self.remaining_length.set(length);
+                    self.buffer_index.set(0);
+
+                    match self.driver.read_page(address / page_size, page_buffer) {
+                        Ok(()) => Ok(()),
+                        Err((error_code, page_buffer)) => {
+                            self.page_buffer.replace(page_buffer);
+                            Err(error_code)
+                        }
+                    }
+                }
+            })
+    }
+
+    fn erase(&self, address: usize, length: usize) -> Result<(), ErrorCode> {
+        if self.state.get() != State::Idle {
+            return Err(ErrorCode::BUSY);
+        }
+
+        self.page_buffer
+            .take()
+            .map_or(Err(ErrorCode::RESERVE), move |page_buffer| {
+                let page_size = page_buffer.as_mut().len();
+
+                self.state.set(State::Erase);
+                self.length.set(length);
+
+                if address % page_size == 0 && length >= page_size {
+                    // This erase is aligned to a page and we are erasing an entire page.
+                    self.address.set(address + page_size);
+                    self.remaining_length.set(length - page_size);
+
+                    self.page_buffer.replace(page_buffer);
+
+                    self.driver.erase_page(address / page_size)
+                } else {
+                    // This erase is non-page-aligned, so we need to do a read first.
+                    self.address.set(address);
+                    self.remaining_length.set(length);
+
+                    match self.driver.read_page(address / page_size, page_buffer) {
```

Comment:
Do we still need to do the erase after this read?

Reply:
This is the non-page-aligned erase scenario.  We do `read-modify-write` to simulate the erase.  It is handled in the block `state:: Erase` within `read_complete`.  However, if it is an actual flash, we might have to add `erase` in the sequence.